### PR TITLE
En 13432 consensus group size changes part 2

### PR DESCRIPTION
**Pull request by bogdan-rosianu**
## Reasoning behind the pull request
- part 2 of the consensus group size changes
  
## Proposed changes
- fix older TODOs
- refactoring & improvements

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: process/rating/ratingsData.go [hunk 1]
```diff
@@ -143,70 +142,91 @@ func NewRatingsData(args RatingsDataArg) (*RatingsData, error) {
...
-	// TODO: add unit tests for the new changes in a future PR
+	// sort the config values descending
 	sort.SliceStable(ratingsStepsConfig, func(i, j int) bool {
-		return ratingsStepsConfig[i].enableEpoch < ratingsStepsConfig[j].enableEpoch
+		return ratingsStepsConfig[i].enableEpoch > ratingsStepsConfig[j].enableEpoch
 	})
 
+	earliestConfig := ratingsStepsConfig[len(ratingsStepsConfig)-1]
+	if earliestConfig.enableEpoch != 0 {
+		return process.ErrMissingConfigurationForEpochZero
```
Comment (iulianpascalau) [inline process/rating/ratingsData.go:167]: :+1:

File: process/rating/ratingsData.go [hunk 2]
```diff
@@ -230,19 +250,16 @@ func (rd *RatingsData) EpochConfirmed(epoch uint32, _ uint64) {
...
-			continue
+		if ratingsStepConfig.enableEpoch <= epoch {
+			return ratingsStepConfig, nil
 		}
-
-		ratingsStep = ratingsStepConfig
 	}
 
-	return ratingsStep
+	// the code should never reach this point, since the config values are checked on the constructor
```
Comment (iulianpascalau) [inline process/rating/ratingsData.go:261]: indeed should never reach this point
it seems like a dead code but that's the way we need to make golang happy :)

File: sharding/chainParametersHolder.go [hunk 3]
```diff
@@ -68,6 +75,18 @@ func validateArgs(args ArgsChainParametersHolder) error {
 		}
 	}
 
+	doesConfigForEpochZeroExist := false
+	for _, chainParams := range chainParametersConfig {
```
Comment (iulianpascalau) [inline sharding/chainParametersHolder.go:79]: this for can be avoided and add a check on L36, just after you sort the configs (same as in ratingsData.go)

File: sharding/chainParametersHolder.go [hunk 4]
```diff
@@ -107,29 +130,31 @@ func (c *chainParametersHolder) AllChainParameters() []config.ChainParametersByE
 	c.mutOperations.RLock()
 	defer c.mutOperations.RUnlock()
 
-	// TODO: return a copy of the object
-	return c.chainParameters
+	chainParametersCopy := make([]config.ChainParametersByEpochConfig, len(c.chainParameters))
+	for idx, chainParameterForEpoch := range c.chainParameters {
```
Comment (iulianpascalau) [inline sharding/chainParametersHolder.go:134]: use `copy(chainParametersCopy, c.chainParameters)` instead ?

File: process/rating/ratingsData_test.go [hunk 5]
```diff
@@ -92,10 +106,34 @@ func createDummyRatingsConfig() config.RatingsConfig {
...
+
+	ratingsDataArg := createDummyRatingsData()
+	ratingsDataArg.ChainParametersHolder = nil
+
+	ratingsData, err := NewRatingsData(ratingsDataArg)
+
+	assert.Nil(t, ratingsData)
+	assert.True(t, errors.Is(err, process.ErrNilChainParametersHandler))
+}
+
```
Comment (iulianpascalau) [inline process/rating/ratingsData_test.go:132]: missing test for no config in epoch 0?

File: sharding/chainParametersHolder.go (hunk 4 above)
Comment (bogdan-rosianu) [reply sharding/chainParametersHolder.go:134]: done. also added test for pointers check

File: process/rating/ratingsData_test.go (hunk 5 above)
Comment (bogdan-rosianu) [reply process/rating/ratingsData_test.go:132]: added

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**